### PR TITLE
Add random_ordered_tree and forest_str

### PR DESCRIPTION
**Pull request by Erotemic**
The purpose of this PR is to simplify the review of #4327. It separates the new self-contained developed for that branch  `random_ordered_tree` which like it sounds make a random ordered tree, and `forest_str`, which makes a text-based representation of a directed or undirected forest. 

After this is finalized and accepted, I'll rebase #4327 on top of master/main. Reviewing these two features should be much easier and make tackling #4327 more feasible. 

This PR ~depends on #4326~ has no dependencies.

This PR is depended on by #4327 and #4350

### CODE REVIEW

**Comment by Erotemic** [comment]
I made a few changes. 

First a bugfix: I noticed that `forest_str` was printing ordered trees in the incorrect order. I fixed this by pushing children onto the stack in reverse order so they are popped in the correct order. 

Second, in the case where `forest_str` is given an undirected graph, there needs to be logic to choose which nodes to use as "source" nodes.  I was hesitant to use `nx.connected_components` because I wanted the function to be able to work an in "eager-evaluation" mode, which is stiffed if there is a linear complexity preprocessing step. However, I realized that determining the sources for the directed case and adding all sources to the stack is also linear complexity, so it really doesn't matter.

Lastly, I changed the name of `use_labels` to `with_labels` to better agree with `nx.draw`.

File: networkx/readwrite/tests/test_text.py
Comment (dschult) [inline networkx/readwrite/tests/test_text.py:37]: Are these print statements intentionally put into the tests so when the tests fail you can actually see what is going on?  That's actually a nice idea. Is it true that pytest never actually shows these print results unless the test fails?

File: networkx/readwrite/tests/test_text.py
Comment (dschult) [inline networkx/readwrite/tests/test_text.py:197]: Should import pytest appear at the module level?
Also, I think we should use `nx.complete_graph` rather than `nx.generators.complete_graph` to reduce maintenance cost of refactoring. Is there an advantage to spelling out the subpackage structure?

**Review by dschult — APPROVED** [review]
This looks good to me.

File: networkx/readwrite/tests/test_text.py
Comment (Erotemic) [reply networkx/readwrite/tests/test_text.py:37]: The print statements serve several purposes. One is for debugging if the test fails. Another was for development. I needed to see the results (after manually verify they were correct of course) so I could populate the expected "want" fields with appropriately formatted text.

Pytest suppresses (doesn't show) output by default, but if you run with `-s` it will disable output suppression. So if you run the tests like this: 

```
pytest networkx/readwrite/tests/test_text.py -v -s
```

You can see the output even if the test does not fail.

File: networkx/readwrite/tests/test_text.py
Comment (Erotemic) [reply networkx/readwrite/tests/test_text.py:197]: Predominately local imports is a side effect of my dev style. Local imports don't bother me especially if they are the only function in the module using the imported module. It makes it easier to cut and paste code between modules. That being said, it may be more pythonic to globally import it in this case, and I don't have a strong opinion on it. 

Me using `nx.generators.complete_graph` was a side effect due to how big the networkx global namespace is. I don't always remember what a function is called, but I knew there was a generators module, so it was easier for me to tab-complete to `complete_graph` after I specified the generators module. Its easy enough to drop the generators piece, and I don't think there is a benefit for specifying it as such given the current structure.

**Comment by Erotemic** [comment]
I was working with this module in other PRs, and there were a few changes that might make using this module easier. 

In networkx.generators I put random_ordered_tree in random_graphs, but perhaps it would be better in the generators.trees module. 

I also thought maybe it would be easier if the `nx.random_tree` function simply took a `create_using` argument (similar to other places in nx). I made a commit that adds that to  `nx.random_tree`. 

Should `random_ordered_tree` even exist if `nx.random_tree` now has a `create_using` kwarg? I left it in for now, but its easy enough to remove. 

I did make one slight functionality change. Previously if you requested a directed graph, it would use dfs starting from an random node as the source. Now, I'm always using the node 0 as a source. It makes the function simpler, and the user can always relabel if they want random labels too.

**Comment by Erotemic** [comment]
After making changes to `nx.random_tree` to include the `create_using` argument, it looks like the doc build step fails because I added a doctest that used utf8 chars. This means that this PR now depends on PR #4326 where I add xetex support to docs. 

This should be merged only after #4326 is merged.

**Comment by Erotemic** [comment]
#4326 was merged.  I rebased on master, so this PR now has no dependencies.

File: networkx/generators/trees.py
Comment (rossbar) [inline networkx/generators/trees.py:241]:
```suggestion
        tree = empty_graph(0, create_using)
```
As I understand it, the `new` function here is to handle the case that users can pass in classes *or* instances to `create_using`. This is hard to parse and I think it would be an improvement if specialized input-handling code could be avoided. Is the suggestion sufficient to accomplish the same behavior?

Side note: this is probably a good example of the motivation for [NXEP 3](https://networkx.org/documentation/latest/developer/nxeps/nxep-0003.html) - a code snippet like this might make a nice addition to the NXEP. CC @boothby

File: networkx/readwrite/text.py
Comment (rossbar) [inline networkx/readwrite/text.py:6]: This function produces nice visualizations, and I think having visualizations that work in the terminal is a really nice feature.

One thing that I've [run into before](https://github.com/numpy/numpy/pull/15666#issuecomment-601337138) is issues with the default fonts for terminals not supporting the full range of unicode characters. This was particularly a problem with Windows, as the default fonts for popular CLI programs only support the Latin1 character range (or some other limited subset). In the linked PR we ended up doing a platform check and adding an ASCII printing style that was the default for windows platforms.

I'm not sure how big of a concern this is here, but I thought I'd mention it as it'd be nice to make sure this is well-supported on all platforms. Unfortunately this is a difficult thing to catch as tests that check the character codes will pass, but the actual character rendering can still fail.

**Review by rossbar — COMMENTED** [review]
I had a few big-picture comments from a quick review here. The PR has some really nice features in it. I had a couple questions/comments about API and platform support. LMK what you think!

File: networkx/generators/random_graphs.py
Comment (dschult) [inline networkx/generators/random_graphs.py:1336]: I think the `@py_random_state` decorator already does this, right?

File: networkx/generators/random_graphs.py
Comment (dschult) [inline networkx/generators/random_graphs.py:1338]: You almost certainly don't need an `OrderedGraph` class here. That would be needed for Python version pre-3.6. But with Python3.6+ the regular classes are ordered because the underlying dict is ordered.

File: networkx/generators/trees.py
Comment (dschult) [inline networkx/generators/trees.py:241]:
```suggestion
        tree = nx.empty_graph(0, create_using)
```

File: networkx/generators/random_graphs.py
Comment (Erotemic) [reply networkx/generators/random_graphs.py:1300]: It can, but I added the latter option towards the end of finalizing this PR. I can remove this function.

File: networkx/generators/trees.py
Comment (Erotemic) [reply networkx/generators/trees.py:241]: I'll have to look into this more later. NXEP seems detailed, and I'd like to read it before I comment in detail.

I will say that the code in "new" exists in other places in networkx. I copied the pattern. I think I was going to propose putting the "new" logic into a classmethod. Perhaps that's addressed in the NXEP or in empty_graph. I'll take a look.

File: networkx/readwrite/text.py
Comment (Erotemic) [reply networkx/readwrite/text.py:6]: I took a lot of time in selecting the characters to use. The ones I've chosen are fairly well supported, but its definitely something we should double check, or perhaps include a fallback mode. I have a bunch of notes on the topic in the comments, I think I removed them in the latest commit, but I'll dig them up.

File: networkx/generators/random_graphs.py
Comment (Erotemic) [reply networkx/generators/random_graphs.py:1336]: I'm personally not a fan of decorators if they don't save you a line. It makes the code easier to statically parse. I can move to the decorator style for the sake of stylistic consistency, but the end result should be functionally equivalent.

Comment (dschult) [reply networkx/generators/random_graphs.py:1336]: But you use the decorator just before the function and then redo what the decorator does. 
Just remove these 3 lines and you're good.

File: networkx/generators/random_graphs.py
Comment (Erotemic) [reply networkx/generators/random_graphs.py:1338]: This function is likely being removed as it is redundant with the new `create_using` arg in `random_tree`.

While dictionaries did becomes ordered by default in 3.6, that guarantee is still not part of the Python language spec AFAIK. I would much prefer to be explicit about the orderedness of graphs. Furthermore, there are mathematical implications of having an ordered graph versus one that has an arbitrary order. The algorithms in the PRs that build on this actually depend on the graph having a defined order for correctness and runtime analysis. I know there are randomized algorithms that depend on random orderings; I'm not sure if any rely on arbitrary orderings.

I'm stressing this to make sure the maintainers know there is value in explicitly using and keeping `Ordered` `Graph/DiGraph/etc...` variants in the API.

File: networkx/generators/trees.py
Comment (Erotemic) [reply networkx/generators/trees.py]: I did accept @dschult's suggestion which is similar to this, but also includes removal of the above comment.

File: networkx/generators/random_graphs.py
Comment (Erotemic) [reply networkx/generators/random_graphs.py]: Ah, I guess I did do that. I actually removed the entire function. I agreed with @rossbar's comment that its best not to expand the API if we are also adding `create_using` to `random_tree`.

File: networkx/generators/random_graphs.py
Comment (dschult) [reply networkx/generators/random_graphs.py]: Thanks for that comment.  
Also, CPython 3.7+ guarantees the order of dicts to be the order keys are added.  I guess if you try to use Pypy that is not guaranteed, but they are still on version2.7 and our current code probably is starting to break on that version.

File: networkx/readwrite/text.py
Comment (Erotemic) [reply networkx/readwrite/text.py:6]: So the characters I chose are "box-drawing characters" (https://en.wikipedia.org/wiki/Box-drawing_character). 

The main characters I use: `╙`, `─`, `╟`, `├` are part of a DOS subset that is widely supported. 

However, I also use two other characters `╎` (to indicate disconnected trees in the same forest) and `╼` (which I use to indicate direction). I'm not sure if they will cause issues. 

I previously used characters from https://stackoverflow.com/questions/2701192/what-characters-can-be-used-for-up-down-triangle-arrow-without-stem-for-displa like `⭬` and `🞂`, which I did run into display issues on the win32 dashboard. So, the  `╎`  and `╼` characters do seem to be more supported than those other arrow unicode chars. 

Do you think we should include a fallback ASCII only mode? I'm not sure how to test for it (especially you're just generating a string and not necessarily sending it to a terminal stream), but we could include a flag `ascii_only=False` the user could explicitly set if they wanted. 

I think the following ascii glyphs would work: 

```
            _write("+")

            if not indent:
                if islast:
                    this_prefix = indent + "L-- "
                    next_prefix = indent + "    "
                else:
                    this_prefix = indent + "|-- "
                    next_prefix = indent + ":   "

            else:
                # For individual forests distinguish between directed and
                # undirected cases
                if is_directed:
                    if islast:
                        this_prefix = indent + "|-> "
                        next_prefix = indent + "    "
                    else:
                        this_prefix = indent + "|-> "
                        next_prefix = indent + "|   "
                else:
                    if islast:
                        this_prefix = indent + "L-- "
                        next_prefix = indent + "    "
                    else:
                        this_prefix = indent + "|-- "
                        next_prefix = indent + "|   "
```

File: networkx/generators/random_graphs.py
Comment (Erotemic) [reply networkx/generators/random_graphs.py]: FWIW pypy has a 2.7 and a 3.6 version. I imagine order will be added to the Python dict spec in the future. My main concern is that networkx maintains the ability to distinguish between a graph that is ordered and a graph that has arbitrary order (even if that's not the implementation under the hood).

File: networkx/readwrite/text.py
Comment (Erotemic) [reply networkx/readwrite/text.py]: I pushed up a commit that did something similar. It looks very ugly, but its not much overhead to add it as a feature. Here are some demo ascii trees: 

Directed: 

```
        +-- 0
        :   |-> 1
        :   |   |-> 3
        :   |   L-> 4
        :   L-> 2
        :       |-> 5
        :       L-> 6
        +── 7
            |-> 8
            |   |-> 10
            |   L-> 11
            L-> 9
                |-> 12
                L-> 13
```

and Undirected:

```
        +-- 0
        :   |-- 1
        :   |   |-- 3
        :   |   L-- 4
        :   L-- 2
        :       |-- 5
        :       L-- 6
        +── 7
            |-- 8
            |   |-- 10
            |   L-- 11
            L-- 9
                |-- 12
                L-- 13
```

File: networkx/generators/random_graphs.py
Comment (jarrodmillman) [reply networkx/generators/random_graphs.py]: We are only testing on PyPy 3.6 now.  I am waiting on this PR https://github.com/actions/setup-python/pull/168 to be merged and I will switch us to PyPy 3.7.  They have been working on getting the PyPy 3.7 support regularly for the last few weeks.

Comment (jarrodmillman) [reply networkx/generators/random_graphs.py]: We have only used ``OrderedGraph`` for testing in the past.  I intended to propose deprecating it once we drop Python 3.6, which will happen very soon.

Comment (dschult) [reply networkx/generators/random_graphs.py]: Be careful about claiming alignment between OrderedGraph and a (mathematical) ordered graph. I don't think it is accurate to use these classes to distinguish between "a graph that is ordered and a graph that has arbitrary order for algorithmic purposes". Certainly it is important to distinguish them mathematically. But both OrderedGraph and Graph have the same deterministic node and edge reporting order based on the order that nodes and edges were added to the object. 

We should deprecate OrderedGraph, though it might be useful to show an example of subclassing.

Comment (Erotemic) [reply networkx/generators/random_graphs.py]: My main concern is that the algorithm I implemented in #4327 explicitly requires an OrderedDiGraph, and I do [check for that](https://github.com/networkx/networkx/blob/9c4f3a59d5e273142d69b6f23b33b687384960b2/networkx/algorithms/minors/tree_embedding.py#L145). But this is just an input type check, and I suppose given that a DiGraph is ordered under the hood, the algorithm will still work. I'm worried that without this check a user might think that they are getting the maximum embedding of unordered graphs (which is an APX-hard problem, and the algo I wrote does not find that). I recognize that functionally they are equivalent. I just believe that the semantics are important for implementation elegance.

That's my main argument for keeping ordered graph variants, and if others don't think its strong enough to justify its existence, then I'll accept that.

Regardless, this function and all references to the OrderedDiGraph and OrderedGraph class no longer exist in this PR. So perhaps we can move forward with merging this and save the deprecation discussion for another issue.

File: networkx/generators/trees.py
Comment (jarrodmillman) [inline networkx/generators/trees.py:189]: Remove this line.  We always assume ``import networkx as nx`` in our doctests.

File: networkx/generators/trees.py
Comment (jarrodmillman) [inline networkx/generators/trees.py:203]: Ditto.

File: networkx/readwrite/text.py
Comment (jarrodmillman) [inline networkx/readwrite/text.py:41]: Ditto.

**Review by jarrodmillman — CHANGES_REQUESTED** [review]
Other than removing the imports from the doctest, this LGTM.  Thanks!

File: networkx/readwrite/text.py
Comment (Erotemic) [reply networkx/readwrite/text.py]: done

**Review by jarrodmillman — APPROVED** [review]

**Pull request merged**
